### PR TITLE
feat: Add UDP socket tracing to enhanced session recording

### PR DESCRIPTION
**Pull request by codingllama**
Add UDP socket tracing, in addition to the already existing TCP support.

UDP sockets don't have to be connected, so instead of tracing connect-like kernel methods, we trace packet sends instead (incoming traffic is not traced for either TCP or UDP).

Because we are tracing sends, this means UDP tracing is inherently more spammy than TCP (audited at the connection level). To alleviate potential audit spam, UDP audit is "rate limited" to a max of one event per "udp_silence_period", with a default of 10s (it, 6 events per socket per minute). See the teleport.yaml changes below.

__Audit changes__

"session.network" events now have an "udp_version" field. Either "version" or "udp_version" are non-zero, depending on whether the connection is TPC or UDP.

Examples (various fields omitted):

```jsonc
{
  "event": "session.network",
  "src_addr": "127.0.0.1",
  "dst_addr": "127.0.0.1",
  "dst_port": 48219,
  "version": 0,
  "operation": 1,
  "udp_version": 4,
  //...
}

{
  "event": "session.network",
  "src_addr": "::1",
  "dst_addr": "::1",
  "dst_port": 44746,
  "version": 0,
  "operation": 1,
  "udp_version": 4,
  //...
}
```

__teleport.yaml changes__

The following toggles are added to enhanced_recording:

```yaml
ssh_service:
  enabled: "yes"
  enhanced_recording:
    enabled: true

    # -- NEW --
    # Enables UDP tracing.
    udp: true

    # -- NEW --
    # Ignore period for spammy sends.
    # At most 1 audit event every 10s. If zeroed all sends are logged.
    udp_silence_period: 10s      

    # -- NEW --
    # Max number of concurrently silenced UDP sockets.
    # Most installs shouldn't care about this.
    udp_silence_buffer_size: 128
```

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
+
+# Vagrantfile is used for BPF testing. See also bpf/README.md.
+#
+# To build Teleport, try to build as much as you can outside of Vagrant and
```
Comment (codingllama) [inline Vagrantfile:6]: The experience here isn't super slick, but it's a load better than nothing and it does capture relevant information, like kernel versions, required dev libraries, where to place libbpf, etc.

**Comment by zmb3** [comment]
`udp_disable_tracing: false` is a weird combination of double negatives. Can we make UDP tracing opt-in and have the schema look like:

```
enhanced_recording:
  enabled: true
  udp: true
```

**Comment by codingllama** [comment]
FYI @zmb3 @russjones, the BPF PR is finally out.

@klizhentas and @xinding33, this adds new user-visible changes to both audit and teleport.yaml. Please take a look at the PR description and let me know your thoughts.

File: bpf/enhancedrecording/network.bpf.c [hunk 2]
```diff
@@ -8,6 +8,9 @@
 
 char LICENSE[] SEC("license") = "Dual BSD/GPL";
 
+// Global toggle for UDP tracing.
+u8 udp_disable SEC(".data") = 1;
```
Comment (codingllama) [inline bpf/enhancedrecording/network.bpf.c:12]: I've gone with disabled by default for the BPF C module (to be extra safe), but enabled by default for the feature itself. Seems like a reasonable choice to me, although we can make the entire feature disabled if that seems safer.

FYI @klizhentas @zmb3

File: lib/bpf/common_test.go [hunk 3]
```diff
@@ -74,49 +76,61 @@ func TestMain(m *testing.M) {
...
+	perfBufferPageCount := defaults.PerfBufferPageCount
+	openPerfBufferPageCount := defaults.OpenPerfBufferPageCount
+	zeroPageCount := 0
+	udpSilencePeriod := defaults.UDPSilencePeriod
+	udpSilenceBufferSize := defaults.UDPSilenceBufferSize
+	customUDPSilencePeriod := 1 * time.Minute
+	customUDPSilenceBufferSize := 42
 
 	var tests = []struct {
-		inConfig  *servicecfg.BPFConfig
```
Comment (codingllama) [inline lib/bpf/common_test.go:83]: Note: this is mostly a refactor of the table so it uses names tests, t.Run and Go idioms (got/want). Tests now diff the entire struct, which catches additions of new defaults.

File: lib/config/fileconf.go [hunk 4]
```diff
@@ -1740,17 +1740,32 @@ type BPF struct {
 
 	// CgroupPath controls where cgroupv2 hierarchy is mounted.
 	CgroupPath string `yaml:"cgroup_path"`
+
+	// UDPDisableTracing disables UDP connection events.
+	UDPDisableTracing string `yaml:"udp_disable_tracing,omitempty"`
```
Comment (zmb3) [inline lib/config/fileconf.go:1745]: Why a string?

File: lib/bpf/network.go [hunk 5]
```diff
@@ -92,6 +109,36 @@ type rawConn6Event struct {
...
+
+func (e *rawConn6Event) SrcIP() net.IP {
+	return ipv6HostToIP(e.SrcAddr)
+}
+
+func (e *rawConn6Event) DstIP() net.IP {
+	return ipv6HostToIP(e.DstAddr)
+}
+
+func ipv4HostToIP(addr uint32) net.IP {
```
Comment (codingllama) [inline lib/bpf/network.go:129]: This and ipv6HostToIP are just moves. At some point I was using the net.IPs in multiple places; not anymore, but I decided to keep the refactor.

File: lib/defaults/defaults.go [hunk 6]
```diff
@@ -528,6 +528,14 @@ const (
 	// Must be power of 2.
 	OpenPerfBufferPageCount = 128
 
+	// UDPSilencePeriod is the default value in which subsequent UDP sends are
+	// silenced to avoid audit noise.
+	UDPSilencePeriod = 10 * time.Second
```
Comment (codingllama) [inline lib/defaults/defaults.go:533]: This default is rather arbitrary, open to suggestions. FYI @zmb3 @rosstimothy.

**Comment by codingllama** [comment]
> `udp_disable_tracing: false` is a weird combination of double negatives. Can we make UDP tracing opt-in and have the schema look like:
> 
> ```
> enhanced_recording:
>   enabled: true
>   udp: true
> ```

You don't really need to set "disable:false", that was more to show the default. That said I like the idea of calling it just "udp". I'll change it and make it opt-in too, instead of opt-out.

File: lib/config/fileconf.go (hunk 4 above)
Comment (codingllama) [reply lib/config/fileconf.go:1745]: This allows for a wider range of values and easy distinction of empty vs false. Also follows the other Enabled field in the same config.

**Comment by codingllama** [comment]
I think the Buf breaking check is reversed? Let me check.

**Comment by codingllama** [comment]
Oh, I forgot to mention, but this is branched from #31891 (adds .vagrant to .gitignore).

**Comment by zmb3** [comment]
> You don't really need to set "disable:false", that was more to show the default. 

Right, I just think it's more natural to set a flag true to enable a feature, as opposed to setting a flag true to disable a feature.

**Comment by codingllama** [comment]
PTAL, added an UDP tracing test.

File: lib/bpf/bpf_test.go [hunk 7]
```diff
@@ -436,6 +548,107 @@ func TestRootBPFCounter(t *testing.T) {
 	counter.Close()
 }
 
+func TestBPF_udpEvents(t *testing.T) {
```
Comment (codingllama) [inline lib/bpf/bpf_test.go:551]: @jakule, I noticed that all other tests are currently skipped. It's not clear to me what the flakiness is. Should I skip this one too?

File: bpf/enhancedrecording/network.bpf.c [hunk 8]
```diff
@@ -124,7 +135,7 @@ int BPF_KPROBE(kprobe__tcp_v4_connect, struct sock *sk)
 SEC("kretprobe/tcp_v4_connect")
 int kretprobe__tcp_v4_connect(struct pt_regs *ctx)
 {
-    return trace_connect_return(PT_REGS_RC(ctx), 4);
+    return trace_connect_return(PT_REGS_RC(ctx), /*ipver=*/4);
```
[the remark this answers is not in the document]
Comment (codingllama) [reply bpf/enhancedrecording/network.bpf.c:138]: Done.

File: lib/bpf/bpf_test.go [hunk 9]
```diff
@@ -436,6 +548,107 @@ func TestRootBPFCounter(t *testing.T) {
...
+	addr := pc.LocalAddr().String()
+
+	send := func(t *testing.T, network, addr string) {
+		runCmd(t, connTrace, networkInCgroupSend, network, addr)
+	}
+
+	receive := func(t *testing.T) *simplifiedEvent {
+		t.Helper()
+		select {
+		case <-time.After(1 * time.Second):
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/bpf/bpf_test.go:621]: Good Q. I can bump it a bit.

**Review by zmb3 — APPROVED** [review]

File: lib/bpf/bpf_test.go [hunk 10]
```diff
@@ -436,6 +548,107 @@ func TestRootBPFCounter(t *testing.T) {
...
+				}
+			case <-ctx.Done():
+				t.Log("Event loop exited")
+				return
+			}
+		}
+	}()
+
+	// "Bind" a random port for our sends. We don't have to receive.
+	pc, err := net.ListenPacket("udp", ":0")
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/bpf/bpf_test.go]: Done.

File: lib/config/fileconf.go [hunk 11]
```diff
@@ -1740,17 +1740,32 @@ type BPF struct {
...
+	// UDP enables UDP connection events.
+	UDP string `yaml:"udp,omitempty"`
+
+	// UDPSilencePeriod is the period in which subsequent UDP sends are silenced
+	// to avoid audit noise.
+	UDPSilencePeriod *time.Duration `yaml:"udp_silence_period,omitempty"`
+
+	// UDPSilenceBufferSize is the max number of concurrently silenced UDP
+	// sockets.
+	UDPSilenceBufferSize *int `yaml:"udp_silence_buffer_size,omitempty"`
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/config/fileconf.go:1753]: I think most configs would just say "udp: true", which a bit shorter, but no strong opinions on my part here.

**Review by rosstimothy — APPROVED** [review]

File: lib/config/fileconf.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/config/fileconf.go:1753]: Asked product to take a look.

**Comment by codingllama** [comment]
After chatting with product we decided to hold this feature for now. I'll see about pulling some of the refactors added here to a separate PR.

Thanks everyone for the reviews and sorry for the noise.

**Pull request closed without merging**